File: DONGUK/프로그래머스_WEEK 1/문자열 압축.py
def solution(s):
    
    N = len(s)
    answer = N
    
    if N == 1:
        return 1
    
    

    for i in range(1, (N//2) + 1):
        cmp = s[:i]
        res = ''
        cnt = 1
        
        for j in range(i, N, i):
            if s[j:j+i] == cmp:
                cnt += 1
                
            else:
                if cnt == 1:
                    res += cmp
                
                else:
                    res += str(cnt) + cmp
                    
                cnt = 1
                cmp = s[j:j+i]
            
            
        if cnt == 1:
            res += cmp

        else:
            res = str(cnt) + res + cmp
            
            

        answer = min(answer, len(res))
        
    return answer

File: KIJUNG/PRO_WEEK_01/문자열압축.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

#define FOR(i,a,b) for(int i=a;i<b;i++)

using namespace std;

int compress(string str, int len){
    int cnt = 0;
    int result = str.size();
    FOR(i,1,str.size()/len){
        if(str.substr((i-1)*len, len) == str.substr(i*len,len)){
            result -= len;
            cnt++;
            if(cnt == 1) result++;
            else if(cnt == 9) result ++;
            else if(cnt == 99) result ++;
            else if(cnt == 999) result ++;

        }
        else {
            cnt = 0;
        }
    }
    return result;
}


int solution(string s) {
    int answer = s.size();
    FOR(i,1,s.size()/2+1){
        answer = min(compress(s,i), answer);
    }
    
    return answer;
}

int main(void){
    cout << solution("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa") << "\n";

    return 0;
}